### PR TITLE
Cpu strided complex unaryop support

### PR DESCRIPTION
**Pull request by dylanbespalko**
Fixed the CI failures in #25850. There was a minor problem with the `DL_RUNTIME_BUG` in `vml.h`.

In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)

Changes in this PR.

- [x]  Added complex support of UnaryOp kernels.

### CODE REVIEW

**Comment by dylanbespalko** [comment]
@ezyang,

I fixed the CI failures in #25850. The DL_RUNTIME_BUG in vml.h needed to be modified for when `std:op` is called on complex numbers. Once the patch of glibc is no longer needed this can be removed. I think these changes are ready for merging.

Thanks,

Dylan

**Comment by dylanbespalko** [comment]
@ezyang 

Are you able to approve these changes?

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: For the record (repeating from the previous PR), we should not have these headers duplicating implementations. I'll make a call on whether or not this should block this PR after reading through the duplication.

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: I'm not sure what the correct number here should be, but the fact that you've written 2 here and below is a huge smell. In the base implementation, the size is `32 / sizeof(T)`. `sizeof(std::complex<double>)` is 16, so I expect 2 here (good!) but 4 in the definition for `std::complex<float>`

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: This is "simplified" from the loop (and thus diverges from base), but I don't really see the utility of unrolling it as you've done here; since `size()` is constexpr won't it get unrolled anyway?

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: It probably would have been ok to replace the `values[i] < 0 ? -values[i] : values[i]` in base with abs, you get the same code in each case (and then the complex overload works): https://godbolt.org/z/UjshUJ

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: ...really?

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (dylanbespalko) [reply aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: Yeah, I appreciate the feedback on that. 

I haven't thought very much about the implementations in vec256_complex... yet.  I just put stuff there so I would stop copy and pasting the kernels for complex ops. I really need to take a fresh look at these implementations. Does that sound ok?

File: aten/src/ATen/native/UnaryOps.cpp
Comment (ezyang) [inline aten/src/ATen/native/UnaryOps.cpp]: 👍

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (dylanbespalko) [reply aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: Thanks for introducing me to this website! The fact that you can show the assembly is the same for both approaches is amazingly helpful.

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [reply aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: Yes, I think taking a closer look is a good idea. I (and others) can answer questions about what the classes are doing. Based on my code review, I think you can directly reuse base for complex, with two additions:

1. You add a way of disabling methods which don't work for complex
2. We do some very minor adjustments to a few functions (like `abs`) to make them more complex friendly

If you need to unblock, I'm even willing to accept a copy of base for complex (in the short term). But I'm not willing to have two copies for float and double.

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (dylanbespalko) [reply aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: I know, these are the easy functions to vectorize. I was planning to do that next week.

**Review by ezyang — CHANGES_REQUESTED** [review]
Most of everything is good, but I don't want to take the duplicate Vec256. If you want to split the patch again without Vec256, I can merge the non-Vec256 bits first.

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (ezyang) [reply aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: I am actually talking more semantics: should `ceil` on complex really be elementwise? (I don't know the answer to this question!)

**Comment by dylanbespalko** [comment]
> Most of everything is good, but I don't want to take the duplicate Vec256. If you want to split the patch again without Vec256, I can merge the non-Vec256 bits first.

Ok, no problem. Just to clarify, I should submit the PR without:
`aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h`
`aten/src/ATen/cpu/vec256/vec256_complex_float_no_vec.h`

But I should leave the other Vec256 changes (real/imag) in?

File: aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h
Comment (dylanbespalko) [reply aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h]: Yes, I would have to profile a couple approaches to see which is better. I agree that vec256_complex_double_no_vec.h and vec256_complex_float_no_vec.h need lots of work before they should be put in.

**Comment by ezyang** [comment]
Yeah. I think you an add real/imag into base and with the templates you've defined so far this will work everywhere.

**Pull request closed without merging**
